pallets/jobs/src/functions.rs

```diff
@@ -121,160 +121,224 @@ impl<T: Config> Pallet<T> {
-					// - Extract information from 'phase1'
-					// - Create a new 'job_type' of ZkSaasPhaseOneJobType with adjusted parameters
-					//   (remove the reported validator)
-					// - Charge the validator fee for job submission
-					// - Store information about the submitted job in 'SubmittedJobs'
-					RoleType::ZkSaaS(role) => {
-						let new_participants = phase1
-							.participants()
-							.ok_or(Error::<T>::InvalidJobPhase)?
-							.into_iter()
-							.filter(|x| x != &validator)
-							.collect::<Vec<_>>()
-							.try_into()
-							.map_err(|_| Error::<T>::TooManyParticipants)?;
-						let phase_one_id = job_info
-							.job_type
-							.get_phase_one_id()
-							.ok_or(Error::<T>::PhaseOneResultNotFound)?;
-						let phase_one = SubmittedJobs::<T>::get(role_type, phase_one_id)
-							.ok_or(Error::<T>::JobNotFound)?;
-						let system = match phase_one.job_type {
-							JobType::ZkSaaSPhaseOne(ref info) => info.system.clone(),
-							_ => return Err(Error::<T>::JobNotFound.into()),
-						};
-
-						let job_type = JobType::ZkSaaSPhaseOne(ZkSaaSPhaseOneJobType {
-							role_type: role,
-							participants: new_participants,
-							system,
-							permitted_caller: phase1.clone().permitted_caller,
+						Self::deposit_event(Event::JobParticipantsUpdated {
+							job_id,
+							role_type,
+							details: job_info.clone(),
 						});
 
-						// charge the validator fee for job submission
-						let job = JobSubmissionOf::<T> {
-							expiry: job_info.expiry,
-							ttl: job_info.ttl,
-							job_type: job_type.clone(),
+						Ok(())
+					} else {
+						let phase1_result = if !job_info.job_type.is_phase_one() {
+							Some(
+								KnownResults::<T>::take(
+									role_type,
+									job_info
+										.job_type
+										.get_phase_one_id()
+										.ok_or(Error::<T>::InvalidJobPhase)?,
+								)
+								.ok_or(Error::<T>::PhaseOneResultNotFound)?,
+							)
+						} else {
+							None
 						};
 
-						let fee = T::JobToFee::job_to_fee(&job);
-						T::Currency::transfer(
-							&validator,
-							&Self::rewards_account_id(),
-							fee,
-							ExistenceRequirement::KeepAlive,
-						)?;
-
-						let job_info = JobInfo {
-							owner: phase1.owner.clone(),
-							expiry: job_info.expiry,
-							ttl: job_info.ttl,
-							job_type,
-							fee,
+						// This phase1 result cannot be used
+						let phase1 =
+							phase1_result.as_ref().ok_or(Error::<T>::PhaseOneResultNotFound)?;
+
+						// Generate a job to generate new key
+						let job_id = Self::get_next_job_id()?;
+
+						match role_type {
+							// Case for RoleType::Tss
+							// - Extract information from 'phase1'
+							// - Create a new 'job_type' of DKGJobType with adjusted parameters
+							//   (remove the reported validator and reduce threshold)
+							// - Charge the validator fee for job submission
+							// - Store information about the submitted job in 'SubmittedJobs'
+							RoleType::Tss(role) => {
```

Comment:
Thoughts on refactoring these matches and break out into helper functions for each role?